teps/0093-add-sign-verify-subcommand-to-the-cli.md

```diff
@@ -0,0 +1,161 @@
+- [Implementation Plan](#implementation-plan)
+  - [Test Plan](#test-plan)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This proposal aims at adding `sign` and `verify` to Tekton CLI to help Trusted Resources users to sign and verify Tekton Resources.
+
+
+## Motivation
+
+There is a growing open-source ecosystem around tekton [catalog](https://github.com/tektoncd/catalog) and [artifactHUB](https://artifacthub.io), where number of tasks/pipelines are made available ready-to-use in our CI/CD. These tasks provides commonly used functions like `git-clone`, `curl`, etc. And also includes vendor specific functions like `aws-cli`, `openshift-client`. Such a marketplace is essential in practicing DRY principle. But, as we are building such an open-distribution framework, we need to ensure task distribution is secure.
+
+Objective in this proposal is to faciliate an easy and handy approach for task/pipeline providers to sign their files and for consumers to verify them using native `tkn cli` commands.
+
+Also, this complements [TEP-0091](https://github.com/tektoncd/community/blob/main/teps/0091-trusted-resources.md). TEP-0091 proposes to apply verification at the controller and the signing is suggested to be implemented into Tekton CLI. This proposal suggests an CLI extension for developers to sign and verify tasks/pipelines YAML files.
+
+### Goals
+
+1. Add the ability to run `tkn {object} sign` and `tkn {object} verify` fom the CLI. `object` currently should support `task` and `pipeline`. `Sign` will take the private key and the resource yaml file, sign the resource object and attach signature to the annotation.
+2. Support both key files and KMS key.
+3. Support v1beta1 Task and Pipeline
+
+### Non-Goals
+
+1. Keyless signing could be a future work
+2. Integrate with Fulcio and Rekor could be future work
+3. Store Signature to a remote storage
+4. Sign and verify a list of files could be future work
+5. Support resources other than v1beta1 Task and Pipeline could be future work
+6. Distribution/Discovery of the public keys is out-of-scope and assumes it is already available for users
+
+### Use Cases
+
+**Catalog Users**: Catalog users consist of Tasks/Pipeline authors and users. For authors they can use tkn to sign their files, distribute and expose the public key to other users who want to use and verify. Other users can use tkn to verify the resources are not tampered with since they are signed by authors.
+
+**Trusted Resources Users**: [TEP-0091](https://github.com/tektoncd/community/blob/main/teps/0091-trusted-resources.md) introduces the verification at pipeline controller but doesn't cover the signing. This tep can provide a tool for Trusted Resources Users to sign their resources.
+
+### Requirements
+
+* Signing a resource shouldn't change the content of it, e.g. we should not call `Setdefaults` for these resources to add default values.
```

Comment:
How is "content" defined? Does it mean anything under `spec`? 
I imaging that the signing operation will add an annotation to the resource.

Reply:
Sure! I will add some clarification to this. The `content` here refers the whole yaml file. e.g., for a task yaml file it will include typemeta, objectmeta data as well as the spec. The signature for now is attached to the annotation, which means that during verification we need to extract, decode and delete the signature annotation before verification.

The future work should include remote signature. It is not considered now to ease the process of implementation. So we don't have to implement signature resolver

Reply:
I think it should be any content (spec, typemeta, objectmeta) except the signature annotation.

Reply:
I will clarify this.